### ISSUE TITLE
Possibly unintended `m_.clear();` in parser ctor?

### ISSUE THREAD

**Issue by DisableAsync** [boostorg/beast#2423]
[Code of question](https://github.com/boostorg/beast/blob/90e37ae40bc80c99d1fae6078a6843a7e1d8f067/include/boost/beast/http/impl/parser.hpp#L36)

I don't understand this ctor overload. We pass stuff in to construct a `http::message`, then we just clear it?
Am I missing something or is this behavior unintended?

**Comment by sehe** [comment]
The parser is used to create a new message from serialized form.

The reason why a reference to a message instance can be passed is to allow re-use (reducing allocations).

The clear makes sure that the output reflects the result of parsing.

It is also assumed for internal logic (e.g. when protocol specifications are verified, lingering message state such as headers might trigger an error-condition or - worse - silently accept a malformed message).

**Comment by kcbsbo** [comment]
well, but it's not reference type only

**Comment by sehe** [comment]
It's not a reference type. Is there a question?

The parser takes ownership of the resources (by move) and the caller can get it back by move via [`release()`](https://www.boost.org/doc/libs/master/libs/beast/doc/html/beast/ref/boost__beast__http__parser/release.html).

**Comment by DisableAsync** [comment]
> The reason why a reference to a message instance can be passed is to allow re-use (reducing allocations).
> 
> The clear makes sure that the output reflects the result of parsing.

Maybe we should update the code comments a bit?

**Comment by DisableAsync** [comment]
This is just so confusing for me.

The [comment of this ctor](https://github.com/boostorg/beast/blob/90e37ae40bc80c99d1fae6078a6843a7e1d8f067/include/boost/beast/http/parser.hpp#L98) says optional arguments are forwarded to message ctor.  Which is same for [this ctor](https://github.com/boostorg/beast/blob/90e37ae40bc80c99d1fae6078a6843a7e1d8f067/include/boost/beast/http/parser.hpp#L140) taking a rvalue reference parser as first parameter. But the latter does not clear message after construction.

**Comment by sehe** [comment]
@DisableAsync The confusion is understandable.

The second constructor is highly specific: it allows a parser to change its body type on the fly, explicitly retaining existing state. Note the extra requirements that are in place:

> The constructed-from parser must not have any parsed body octets or initialized <em>BodyReader</em>, otherwise an exception is generated.

This is explained in the documentation as "Deferred Body Type" in [Change Body Type 💡](https://www.boost.org/doc/libs/1_79_0/libs/beast/doc/html/beast/more_examples/change_body_type.html).

In other words, the second constructor is a highly specific use-case assuming you Know What You're Doing. The first is necessarily more cautious to prevent issues like issue #818 (see also https://github.com/boostorg/beast/blob/develop/test/beast/http/parser.cpp#L332)

**Comment by DisableAsync** [comment]
> The first is necessarily more cautious to prevent issues like issue #818 (see also https://github.com/boostorg/beast/blob/develop/test/beast/http/parser.cpp#L332)

This issue looks more like a library API misuse (reusing the `request` object for different requests without clearing its states) to me.

> The second constructor is highly specific: it allows a parser to change its body type on the fly, explicitly retaining existing state. Note the extra requirements that are in place:
>> The constructed-from parser must not have any parsed body octets or initialized BodyReader, otherwise an exception is generated.

I suppose that means the only state we actually reuse is the message header?

If one can construct a new parser from an existing one, so that they get to use the constructed-from parser's header data., then why can't they just pass the header data itself to the parser for reuse?

**Comment by DisableAsync** [comment]
> If you can describe a valid use-case not supported by the existing functionality, ...

In that case, nope. 

Maybe I'm just overthinking too much. :P
